### PR TITLE
Stop Expression.TypeEquals throwing on Nullable<T> parameter.

### LINKED ISSUE #5272: Expression.TypeEquals throws on Nullable<T> parameter.
```C#
Expression value = Expression.Constant(DateTime.MaxValue, typeof(DateTime?));		
var param = Expression.Parameter(value.Type);

Func<bool> func = Expression.Lambda<Func<bool>>(
	Expression.Block(
		new[] { param },
		Expression.Assign(param, value),
		Expression.TypeEqual(param, typeof(object))
	)
).Compile();
```

The above throws `InvalidOperationException` on the `Compile()` with both compilation options. This happens for any type for the `type` parameter except for the actual nullable type of the parameter which correctly returns `false` if the value of `value` is null and `true` otherwise. This includes the related non-nullable type (should return the same as with the nullable type).

### PR DESCRIPTION
**Pull request by JonHanna**
Fixes #5272

### CODE REVIEW

**Comment by JonHanna** [comment]
Test this please.

**Comment by JonHanna** [comment]
Test Innerloop CentOS7.1 Debug Build and Test
Test Innerloop CentOS7.1 Release Build and Test
Test Innerloop Ubuntu Debug Build and Test
Test Innerloop Ubuntu Release Build and Test

**Comment by JonHanna** [comment]
Test Innerloop CentOS7.1 Debug Build and Test
Test Innerloop CentOS7.1 Release Build and Test
Test Innerloop Ubuntu Debug Build and Test
Tes Innerloop Ubuntu Release Build and Test

**Comment by JonHanna** [comment]
Innerloop Ubuntu Release Build and Test

**Comment by JonHanna** [comment]
Test Innerloop Ubuntu Release Build and Test

**Comment by JonHanna** [comment]
Test Innerloop CentOS7.1 Release Build and Test
Test Innerloop CentOS7.1 Debug Build and Test 
Test Innerloop Ubuntu Debug Build and Test

**Comment by JonHanna** [comment]
#5279 has tests relevant to this. Currently avoiding the data which triggers it, commented so that they can be changed to hit the relevant cases.

**Comment by JonHanna** [comment]
#5279 has tests relevant to this. Currently avoiding the data which triggers it, commented so that they can be changed to hit the relevant cases.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/TypeBinaryExpression.cs [hunk 1]
```diff
@@ -79,11 +79,17 @@ internal Expression ReduceTypeEqual()
...
+            // If the expression type is a sealed reference type or a nullable
+            // type, it will match if the value is not null and the type operand
+            // either matches or one is a nullable type while the other is its
+            // type argument (T to the other's T?).
+            if (cType.GetTypeInfo().IsSealed)
             {
-                if (cType.IsNullableType())
+                if (cType.GetNonNullableType() != _typeOperand.GetNonNullableType())
+                {
+                    return Expression.Constant(false);
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/TypeBinaryExpression.cs]: It seems we're no longer evaluating `Expression` for its side-effects in the cases where we return a constant now.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/TypeBinaryExpression.cs [hunk 2]
```diff
@@ -93,6 +99,13 @@ internal Expression ReduceTypeEqual()
                 }
             }
 
+            // If the type operand is sealed (including value types, void and nullables) then the result
+            // must be false or it would have been handled by one of the cases above.
+            if (_typeOperand.GetTypeInfo().IsSealed)
+            {
+                return Expression.Constant(false);
```
Comment (bartdesmet) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/TypeBinaryExpression.cs]: Same as the remark above.

File: src/System.Linq.Expressions/src/System/Linq/Expressions/TypeBinaryExpression.cs (hunk 1 above)
Comment (JonHanna) [inline src/System.Linq.Expressions/src/System/Linq/Expressions/TypeBinaryExpression.cs]: Right you are. Will fix that promptly.

**Comment by JonHanna** [comment]
Test this, please

**Comment by VSadov** [comment]
@JonHanna - do you plan to enable tests form #5279 in this change?
conversion to object seems indeed be unreachable. Perhaps replace with assert?

**Comment by JonHanna** [comment]
@VSadov I'm pulling down the current master to do exactly that right now.

**Comment by JonHanna** [comment]
@VSadov that's the first of three with the tests merged in :)

And yes, I put in an assert, and before what had been prematurely optimising parameter expressions.

**Comment by VSadov** [comment]
LGTM

**Pull request merged**